### PR TITLE
Add experimental support for DANE / TLSA

### LINKED ISSUE #784: Add support for DANE / TLSA
ZNC should try to authenticate certificates using DANE when connecting to TLS servers.

[DANE](https://en.wikipedia.org/wiki/DNS-based_Authentication_of_Named_Entities) is a protocol which replaces or reinforces the X.509 validation of certificates used by SSL/TLS, which is usually done by checking the signature added by a certification authority. With DANE, the certificate fingerprint can also be checked with a DNS query, hopefully authenticated by DNSSEC. It is very useful with self-signed certificates, as it allows the server administrator to explicitly tell the client which certificate it should receive and to ignore the signature validation.

References:
- [RFC 6698](http://tools.ietf.org/html/rfc6698)
- [irssi's announcement](http://ahf.me/articles/2013/09/14/enhancing-ssl-security-for-irc-dane-support/) with a more detailed description and motivation
- [irssi's implementation](https://github.com/irssi/irssi/blob/8bd575df2ec0d4a17b6f59116d555b64f5bb1312/src/core/network-openssl.c#L209-L241) which uses the [dnsval library](http://dnssec-tools.org/) (libval-{bin,dev} on Debian Jessie)
## 

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/7182050-add-support-for-dane-tlsa?utm_campaign=plugin&utm_content=tracker%2F1759&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F1759&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Alef-Burzmali**
This commit adds experimental support for DANE validation of TLS server certificates with the libval library from the DNSSEC-Tools package, like the [irssi implementation](https://github.com/irssi/irssi/blob/8bd575df2ec0d4a17b6f59116d555b64f5bb1312/src/core/network-openssl.c#L209-L241).

You have to [patch Csocket](https://gist.github.com/Alef-Burzmali/d6ebbf6d57a6f4fab191) to use it, as the library requires the `ssl` object to test the certificate(s) directly. I'll open a PR there if this one is well received.

I've tested the `PKIX-EE(1)` and `DANE-EE(3)` usages with both valid and invalid records, I was unable to test `PKIX-TA(0)` or `DANE-TA(2)`. `DANE-EE` should be the most common usage anyway, it's the one that validates self-signed certs.

As this is the biggest C++ patch I've ever written, I'll appreciate any feedback. :-)

Fix #784.

### CODE REVIEW

**Comment by Mikaela** [comment]
> You have to patch Csocket to use it, as the library requires the ssl object to test the certificate(s) directly. I'll open a PR there if this one is well received.

I highly doubt this is merged until there is PR at Csocket.

https://github.com/jimloco/Csocket

File: src/Socket.cpp [hunk 1]
```diff
@@ -103,6 +108,40 @@ void CZNCSock::SSLHandShakeFinished() {
...
 		return;
 	}
+
+#ifdef HAVE_DANE
+	struct val_danestatus* pDaneStatus = NULL;
+	struct val_daneparams pDaneParams;
+	pDaneParams.port = GetPort();
+	pDaneParams.proto = DANE_PARAM_PROTO_TCP;
+
+	if (val_getdaneinfo(NULL, m_HostToVerifySSL.c_str(), &pDaneParams, &pDaneStatus) == VAL_DANE_NOERROR) {
```
Comment (DarthGandalf) [inline src/Socket.cpp]: This will block ZNC until DNS query is done.

**Comment by Alef-Burzmali** [comment]
Ok, I've put the DNS query in another thread if HAVE_THREADED_DNS is defined. However, as it is the first time I use threads in C++, I would really appreciate if someone could verify I did not do something stupid.

I've also tested `PKIX-TA(0)` or `DANE-TA(2)`, which work.

**Comment by DarthGandalf** [comment]
Right now ZNC creates 2 jobs in separate threads for sockets: DNS resolving of target host, and DNS resolving of bind host. Maybe this should just be a third one?

**Comment by Alef-Burzmali** [comment]
Yes, it should be easier :)

Thanks @psychon for your feedback. You were right, there was an issue with the destructor. With @DarthGandalf's suggestion, it's gone and there no need for CThreadPool::waitForJob() either.

File: src/Socket.cpp [hunk 2]
```diff
@@ -51,6 +60,10 @@ CZNCSock::CZNCSock(int timeout) : Csock(timeout) {
 }
 
 CZNCSock::CZNCSock(const CString& sHost, u_short port, int timeout) : Csock(sHost, port, timeout) {
+#ifdef HAVE_DANE
+	m_iDaneLookupResult = VAL_DANE_INTERNAL_ERROR;
+	m_pDaneStatus = NULL;
+#endif
```
Comment (DarthGandalf) [inline src/Socket.cpp]: Probably it's time to move this from both constructors to some common function.

**Comment by DarthGandalf** [comment]
@ahf you're more familiar with DANE than me. Does this look fine?

**Comment by Alef-Burzmali** [comment]
Rebased on top of master and replaced NULL by nullptr.

**Comment by Alef-Burzmali** [comment]
Rebased on top of master.
I've try to comply with the new coding style, but I may have forgotten some bits, please review it.

File: src/Socket.cpp [hunk 3]
```diff
@@ -93,6 +128,67 @@ int CZNCSock::VerifyPeerCertificate(int iPreVerify, X509_STORE_CTX * pStoreCTX)
...
+	m_pDaneStatus = pDaneStatus;
+}
+
+bool CZNCSock::VerifyDaneAuthentication(SSL* pSSL, bool bSSLHostOk) {
+#ifndef HAVE_THREADED_DNS
+	u_short iPort = GetPort();
+
+	DEBUG(GetSockName() + ": DANE: initiating TLSA lookup for [" << m_HostToVerifySSL << "], port [" << iPort << "]");
+
+	struct val_daneparams pDaneParams;
```
Comment (DarthGandalf) [inline src/Socket.cpp]: "p" stands for pointer, this is not

File: src/Socket.cpp [hunk 4]
```diff
@@ -93,6 +128,67 @@ int CZNCSock::VerifyPeerCertificate(int iPreVerify, X509_STORE_CTX * pStoreCTX)
...
+	m_iDaneLookupResult = val_getdaneinfo(nullptr, m_HostToVerifySSL.c_str(), &pDaneParams, &m_pDaneStatus);
+
+	DEBUG(GetSockName() + ": DANE: finished TLSA lookup for [" << m_HostToVerifySSL << "], port [" << iPort << "]");
+#endif /* !HAVE_THREADED_DNS */
+
+	if (m_iDaneLookupResult != VAL_DANE_NOERROR || m_pDaneStatus == nullptr) {
+		val_free_dane(m_pDaneStatus);
+		m_pDaneStatus = nullptr;
+
+		DEBUG(GetSockName() + ": DANE: No usable TLSA record");
```
Comment (DarthGandalf) [inline src/Socket.cpp]: Let's have different error messages for these cases. Easier to debug.

File: src/Socket.cpp [hunk 5]
```diff
@@ -93,6 +128,67 @@ int CZNCSock::VerifyPeerCertificate(int iPreVerify, X509_STORE_CTX * pStoreCTX)
...
+	int iPkixValidation = 1;
+	int iDaneResult = val_dane_check(nullptr, pSSL, m_pDaneStatus, &iPkixValidation);
+
+	bool bCertVerified = false;
+	if (iDaneResult == VAL_DANE_NOERROR) {
+		// DANE is succesful
+		DEBUG(GetSockName() + ": DANE: Found TLSA record [" << m_pDaneStatus->usage << " " << m_pDaneStatus->selector << " " << m_pDaneStatus->type << "]");
+
+		if (m_pDaneStatus->usage == DANE_USE_TA_ASSERTION && !bSSLHostOk) {
+			DEBUG(GetSockName() + ": DANE: Valid TLSA record, but certificate is not valid for this host and usage is DANE-TA");
```
Comment (DarthGandalf) [inline src/Socket.cpp]: What's the reason for this branch?

**Comment by Alef-Burzmali** [comment]
I've hopefully fixed the 2 issues, and added a comment to explain the reason behind the branch.

However, further testing has shown that the library I've used, libval 2.0, suffers from major flaws that prevent it from being used. The implementation of usage `DANE-TA(2)` is broken and allow to bypass the certificate validation entirely.

Some possible solutions:
* Disable `DANE-TA(2)`:
  * `DANE-EE(3)` is more likely to be used, as it works with any type of certificates, including self-signed ones;
  * `DANE-TA(2)` requires a self-signed CA used to sign several certificates, which is less common; and these certificates need to be valid (not expired, and match the server name);
  * `DANE-TA(2)` can be replaced by `DANE-EE(3)` by listing all the possible certificate in the DNS;
* Use libval 2.1, which is much more secure, but is not yet packaged and has a different API;
* Use another library. Maybe gnutls?
* Wait until DANE support is added to openssl.

I will try to use libval 2.1 instead, in the meantime, this PR should not be used.

**Comment by DarthGandalf** [comment]
Hm.
Good that you've noticed the issue.

**Comment by DarthGandalf** [comment]
Is there an upstream bug, or maybe CVE, about it?

**Comment by Mikaela** [comment]
> * Use another library. Maybe gnutls?

GNUTLS won't work, because it's DANE depends on Unbound which again depends on OpenSSL and Debian and other distributions aren't packaging it with DANE support for this reason.

Via https://github.com/weechat/weechat/pull/121

**Comment by Alef-Burzmali** [comment]
> Is there an upstream bug, or maybe CVE, about it?

I've not found any CVE nor bug report, even though the developers have been [aware](https://www.dnssec-tools.org/pipermail/users/2013-July/000177.html) of the issue since 2013, from [a review](https://www.ietf.org/mail-archive/web/dane/current/msg05758.html) on the DANE working group mailing-list.

However, version 2.0 is no longer supported and has been replaced by [version 2.1](https://dnssec-tools.org/download/#gotodnsval). The main reason I've used it is because it's the version packaged in Debian testing.

**Comment by Mikaela** [comment]
GnuTLS issue on unbound (and OpenSSL or nss) depedency: https://gitlab.com/gnutls/gnutls/issues/21

**Comment by Alef-Burzmali** [comment]
For those of you who were interested, here is the Debian [bug report for libval](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=797470)

**Pull request closed without merging**